sys/kern/kcsan.c

```diff
@@ -0,0 +1,252 @@
+  }
+}
+
+/*
+ * For a given memory access returns the conflicting watchpoint if the data race
+ * is found. Otherwise returns NULL.
+ */
+static inline atomic_int *search_for_race(uintptr_t addr, size_t size,
+                                          bool is_read) {
+  uintptr_t other_addr;
+  size_t other_size;
+  bool other_is_read;
+
+  int slot = watchpoint_slot(addr);
+  atomic_int *watchpoint_p = &watchpoints[slot];
+
+  int encoded = atomic_load(watchpoint_p);
+  if (!decode_watchpoint(encoded, &other_addr, &other_size, &other_is_read))
+    return NULL;
+
+  /*
+   * Data race occurs when there are two memory accesses and at least one of
+   * them is write.
+   */
+  if (other_is_read && is_read)
+    return NULL;
+
+  if (address_match(addr, size, other_addr, other_size))
+    return watchpoint_p;
+
+  return NULL;
+}
+
+static inline atomic_int *insert_watchpoint(uintptr_t addr, size_t size,
+                                            bool is_read) {
+  int slot = watchpoint_slot(addr);
+  atomic_int *watchpoint_p = &watchpoints[slot];
+
+  int encoded = encode_watchpoint(addr, size, is_read);
+  int expected = WATCHPOINT_INVALID;
+  if (!atomic_compare_exchange_strong(watchpoint_p, &expected, encoded))
+    return NULL;
+  return watchpoint_p;
+}
+
+static inline uint64_t read_mem(uintptr_t addr, size_t size) {
+  switch (size) {
+    case 1:
+      return *((uint8_t *)addr);
+    case 2:
+      return *((uint16_t *)addr);
+    case 4:
+      return *((uint32_t *)addr);
+    case 8:
+      return *((uint64_t *)addr);
+    default:
+      return 0;
+  }
+}
+
+static inline void setup_watchpoint(uintptr_t addr, size_t size, bool is_read) {
+  uint64_t prev_val = 0, new_val = 0;
+
+  /* Reset the counter. */
+  atomic_store(&skip_counter, SKIP_COUNT);
+
+  atomic_int *watchpoint_p = insert_watchpoint(addr, size, is_read);
+  if (watchpoint_p == NULL) {
+    atomic_fetch_add(&kcsan_slot_taken_count, 1);
+    return;
+  }
+
+  atomic_fetch_add(&kcsan_set_watchpoint_count, 1);
+
+  prev_val = read_mem(addr, size);
+  delay(WATCHPOINT_DELAY);
+  new_val = read_mem(addr, size);
+
+  if (__predict_false(prev_val != new_val)) {
+    panic("===========KernelConcurrencySanitizer===========\n"
+          "* value of the watched variable %p has changed\n"
+          "* %s of size %lu\n"
+          "* previous value %x\n"
+          "* current value %x\n"
+          "============================================",
+          (void *)addr, (is_read ? "read" : "write"), size, prev_val, new_val);
+  }
+
+  atomic_store(watchpoint_p, WATCHPOINT_INVALID);
+}
+
+static void kcsan_check(uintptr_t addr, size_t size, bool is_read) {
+  if (!kcsan_ready || cpu_intr_disabled() || preempt_disabled() ||
+      size > MAX_ENCODABLE_SIZE)
+    return;
+
+  if (addr < KERNEL_SPACE_BEGIN)
+    return;
+
+  atomic_int *watchpoint = search_for_race(addr, size, is_read);
+  if (__predict_true(watchpoint == NULL)) {
+    if (should_watch())
```

Comment:
`size` isn't guaranteed to be a power of 2 due to `__tsan_{read,write}_range()`. We could either check that it is before setting up a watchpoint, or round it down to the nearest power of 2 (which is what is done currently, since `log2()` rounds down). If we're going to stick with rounding down, there should probably be a comment about it somewhere, maybe in `setup_watchpoint()`.